### DISCUSSION TITLE
Add possibility to enqueue/append requests with server.use instead of only being able to append

### DISCUSSION THREAD

**Discussion by MartinJaskullaAccenture** [mswjs/msw#1176] [Ideas]
This is how I would like to write my unit tests (I wrapped `server.use` in the `mockOnce` utility function):

```ts
mockOnce<MyInterface1>("get", myMockData1);
mockOnce<MyInterface2>("get", myMockData2);
clickAddToBasket()

mockOnce<MyInterface3>("get", myMockData3);
clickGoToCheckout()

expect(hasUncalledMockOnceCalls()).toBe(false);
```

By reading this I would assume that `clickAddToBasket` triggers 2 api calls in the order they appear in the test. However the first api call would return `myMockData2`, because `msw` essentially uses a stack instead of a queue.

It would be nice if server.use would accept options, which allow me to enqueue/append my requests instead of appending them (In theory I would like to enqueue the `server.use` calls, but keep the original handlers passed to `setupServer` last in the queue).

If this is not feasible, then the feature request [Get currently registered handlers](https://github.com/mswjs/msw/discussions/831) would also help with my use case. Then it would be easy to get the current handlers in `mockOnce` where I could do my custom enqueueing.

**Comment by kettanaito** [comment]
Hey, @MartinJaskullaAccenture.

Looks like you're looking for [res.once()](https://mswjs.io/docs/api/response/once). Once a handler uses a mocked response composed via `res.once()`, that handler gets "deactivated" and is ignored for the rest of the response resolution cycle. You can reset its ignored state via [`.restoreHandlers`](https://mswjs.io/docs/api/setup-server/restore-handlers) at any point.

**Reply by MartinJaskullaAccenture** [reply]
Not sure I am following. In my `mockOnce` function I am using `server.use(rest.get((req, res) => res.once))`, but `server.use` prepends the handlers, but I want to append them instead.

**Reply by MartinJaskullaAccenture** [reply]
Without my util function
```ts
server.use(rest.get("*", (req, res, ctx) => res.once(ctx.status(200), ctx.json(myMockData1), ctx.delay(150))))
server.use(rest.get("*", (req, res, ctx) => res.once(ctx.status(200), ctx.json(myMockData2), ctx.delay(150))))
clickAddToBasket()

server.use(rest.get("*", (req, res, ctx) => res.once(ctx.status(200), ctx.json(myMockData3), ctx.delay(150))))
clickGoToCheckout()

expect(hasUncalledMockOnceCalls()).toBe(false);
```

Ah okay I got it.

What I need to do is:

```ts
server.use(
  rest.get(),
  rest.get()
)
``` 
instead of 
```ts
server.use(rest.get())
server.use(rest.get())
``` 
Can't use my concise util function I guess

**Reply by kettanaito** [reply]
I'm afraid the internal order of handlers shouldn't be controlled by the public API. If you append a handler, it has a chance of never taking effect because there may be a preceding handler defined _earlier_ which will match a request. With `server.use()` you declare an intention to customize previously declare behavior. That's why internally it prepends handlers. 

If you want to abstract code, abstract the handler part, not the `server.use`

```js
server.use(
  mockOnce(...),
  mockOnce(...)
)
```
